shared/js/content-scripts/block-cookie.js

```diff
@@ -28,72 +28,109 @@
                 }
             }
-
-            if (!expiryFound) {
-                comps.push('max-age=' + nextWeek)
+            window.addEventListener('message', messageListener)
+        })
+        console.log('inject cookie test', cookieSetter, secret)
+        document.__defineSetter__('cookie', (value) => {
+            // call the native document.cookie implementation. This will set the cookie immediately
+            // if the value is valid. We will override this set later if the policy dictates that
+            // the expiry should be changed.
+            cookieSetter.apply(document, [value])
+            try {
+                // determine the origins of the scripts in the stack
+                const scriptOrigins = (new Error().stack.split('\n')).reduce((origins, line) => {
+                    const res = line.match(lineTest)
+                    if (res && res[2]) {
+                        origins.add(new URL(res[2]).hostname)
+                    }
+                    return origins
+                }, new Set())
+                // wait for config before doing same-site tests
+                loadPolicy.then(({ tabRegisteredDomain, policy }) => {
+                    const sameSiteScript = [...scriptOrigins].every((host) => host === tabRegisteredDomain || host.endsWith(`.${tabRegisteredDomain}`))
+                    if (sameSiteScript) {
+                        // cookies set by scripts loaded on the same site as the site are not modified
+                        console.log('ignored', value, scriptOrigins)
+                        return
+                    }
+                    // extract cookie expiry from cookie string
+                    const cookieParts = value.split(';')
+                    const maxAgeIdx = cookieParts.findIndex(part => part.trim().toLowerCase().startsWith('max-age'))
+                    const expiresIdx = cookieParts.findIndex(part => part.trim().toLowerCase().startsWith('expires'))
+                    if (maxAgeIdx === -1 && expiresIdx === -1) {
+                        // session cookie
+                        return
+                    }
+                    const expiry = maxAgeIdx >= 0
+                        ? parseInt(cookieParts[maxAgeIdx].split('=')[1])
+                        : (new Date(cookieParts[expiresIdx].split('=')[1]) - new Date()) / 1000
+                    // apply cookie policy
+                    if (expiry > policy.threshold) {
+                        if (maxAgeIdx === -1) {
+                            cookieParts.push(`max-age=${policy.maxAge}`)
+                        } else {
+                            cookieParts.splice(maxAgeIdx, 1, `max-age=${policy.maxAge}`)
+                        }
+                        console.log('update cookie', cookieParts.join(';'), scriptOrigins)
+                        cookieSetter.apply(document, [cookieParts.join(';')])
+                    }
```

Comment:
gosh, that cookie code is scary. Wish Cookie Storage API (https://wicg.github.io/cookie-store/) was available in FF. But since it isn't… can we somehow unit test this code? 🤔

Reply:
Again, this code is tricky to extract in a way that we can unittest because of the injection method. However, if we can extract it we could certainly test it by throwing a load of cookie strings see in the wild at it (e.g. from tracker-radar data) to see if there are any missed edge cases.

Reply:
Can we use the `content-scope` injection method I added in my patch? This then keeps the injected code into it's own file which a unit test could load in the same way / import or do whatever without all the other utility code.